drivers/spi/spi_nxp_lpspi/spi_nxp_lpspi_dma.c

```diff
@@ -127,94 +188,124 @@ static inline int lpspi_dma_rxtx_load(const struct device *dev)
-		status = 0;
 	}
 
-	if (channel == rx->channel) {
-		spi_context_update_rx(ctx, 1, rx->dma_blk_cfg.block_size);
-		debug_char = 'R';
-		rx->chunk_done = true;
-	} else if (channel == tx->channel) {
-		spi_context_update_tx(ctx, 1, tx->dma_blk_cfg.block_size);
-		debug_char = 'T';
-		tx->chunk_done = true;
-	} else {
-		/* invalid channel */
-		status = -EIO;
+	if (channel != dma_data->dma_tx.channel && channel != dma_data->dma_rx.channel) {
+		ret = -EIO;
 		goto error;
 	}
 
-	LOG_DBG("DMA %cX Block Complete", debug_char);
-
-	/* wait for the other channel to finish if needed */
-	if (!rx->chunk_done || !tx->chunk_done) {
-		return;
-	}
-
-
-	while ((IS_ENABLED(CONFIG_SOC_FAMILY_NXP_IMXRT) ||
-		IS_ENABLED(CONFIG_SOC_FAMILY_KINETIS)) &&
-		(base->SR & LPSPI_SR_MBF_MASK)) {
-		/* wait until module is idle */
-	}
-
-	if (spi_context_max_continuous_chunk(ctx) == 0) {
-		goto done;
-	}
-
-	status = lpspi_dma_next_fill(spi_dev);
-	if (status) {
-		goto error;
+	switch(dma_data->spi_transfer_state)
+	{
+		case LPSPI_TRANSFER_STATE_ONGOING:
+		{
+			spi_context_update_tx(ctx, 1, tx->dma_blk_cfg.block_size);
+			spi_context_update_rx(ctx, 1, rx->dma_blk_cfg.block_size);
+			/* Calculate next DMA transfer size */
+			dma_data->synchronize_dma_size = spi_context_max_continuous_chunk(ctx);
+
+			LOG_DBG("tx len:%d rx len:%d next dma size:%d",
+				ctx->tx_len, ctx->rx_len, dma_data->synchronize_dma_size);
+
+			if (dma_data->synchronize_dma_size > 0)	{
+				ret = (channel == dma_data->dma_tx.channel) ?
+				lpspi_dma_tx_load(spi_dev, ctx->tx_buf, dma_data->synchronize_dma_size) :
+				lpspi_dma_rx_load(spi_dev, ctx->rx_buf, dma_data->synchronize_dma_size);
+
+				if (ret != 0) {
+					goto error;
+				}
+
+				ret = dma_start(dev, channel);
+				if (ret != 0) {
+					goto error;
+
+				}
+				dma_data->spi_transfer_state = LPSPI_TRANSFER_STATE_NEXT_DMA_SIZE_UPDATED;
+			} else {
+				ret = dma_stop(dev, channel);
+				if (ret != 0) {
+					goto error;
+				}
+				/* This is the end of the transfer. */
+				if (channel == dma_data->dma_tx.channel) {
+					spi_mcux_issue_TCR(spi_dev);
+					dma_data->spi_transfer_state = LPSPI_TRANSFER_STATE_TX_DONE;
+					base->DER &= ~LPSPI_DER_TDDE_MASK;
+				} else {
+					dma_data->spi_transfer_state = LPSPI_TRANSFER_STATE_RX_DONE;
+					base->DER &= ~LPSPI_DER_RDDE_MASK;
+				}
+			}
+			break;
+		}
+		case LPSPI_TRANSFER_STATE_NEXT_DMA_SIZE_UPDATED:
+		{
+			ret = (channel == dma_data->dma_tx.channel) ?
+			lpspi_dma_tx_load(spi_dev, ctx->tx_buf, dma_data->synchronize_dma_size) :
+			lpspi_dma_rx_load(spi_dev, ctx->rx_buf, dma_data->synchronize_dma_size);
+			dma_data->synchronize_dma_size = 0;
+
+			if (ret != 0) {
+				goto error;
+			}
+
+			ret = dma_start(dev, channel);
+			if (ret != 0) {
+				goto error;
+
+			}
+
+			dma_data->spi_transfer_state = LPSPI_TRANSFER_STATE_ONGOING;
+			break;
+		}
+		case LPSPI_TRANSFER_STATE_TX_DONE:
+		case LPSPI_TRANSFER_STATE_RX_DONE:
```

Comment:
it seems like there should be one state called like BOTH_DONE

Reply:
yes, we can add both_done, but it is meaningless because it is same with NULL state. No place will use it.